plugins/blockvolume/blockprovider/gluster-block/glusterblock.go

```diff
@@ -0,0 +1,272 @@
+	if err != nil {
+		return nil, err
+	}
+	gb.client = gbClient
+
+	return gb, nil
+}
+
+// CreateBlockVolume will create a gluster block volume with given name and size having `hostVolume` as hosting volume
+func (g *GlusterBlock) CreateBlockVolume(name string, size int64, hosts []string, hostVolume string, options ...blockprovider.BlockVolOption) (blockprovider.BlockVolume, error) {
+	var (
+		blockVolOpts = &blockprovider.BlockVolumeOptions{}
+		clusterLocks = transaction.Locks{}
+	)
+
+	blockVolOpts.ApplyOpts(options...)
+
+	req := &api.BlockVolumeCreateReq{
+		HaCount:            blockVolOpts.Ha,
+		AuthEnabled:        blockVolOpts.Auth,
+		FullPrealloc:       blockVolOpts.FullPrealloc,
+		Size:               uint64(size),
+		Storage:            blockVolOpts.Storage,
+		RingBufferSizeInMB: blockVolOpts.RingBufferSizeInMB,
+		Hosts:              hosts,
+	}
+
+	volInfo, err := volume.GetVolume(hostVolume)
+	if err != nil {
+		return nil, fmt.Errorf("error in getting host vol details: %s", err)
+	}
+
+	if err := clusterLocks.Lock(volInfo.Name); err != nil {
+		log.WithError(err).Error("error in acquiring cluster lock")
+		return nil, err
+	}
+
+	defer clusterLocks.UnLock(context.Background())
+
+	availableSizeInBytes, err := strconv.Atoi(volInfo.Metadata["block-hosting-available-size"])
+	if err != nil {
+		return nil, err
+	}
+
+	if int64(availableSizeInBytes) < size {
+		return nil, fmt.Errorf("available size is less than requested size,request size: %d, available size: %d", size, availableSizeInBytes)
+	}
+
+	resp, err := g.client.CreateBlockVolume(hostVolume, name, req)
+	if err != nil {
+		return nil, err
+	}
+
+	volInfo.Metadata["block-hosting-available-size"] = fmt.Sprintf("%d", int64(availableSizeInBytes)-size)
+
+	if err := volume.AddOrUpdateVolume(volInfo); err != nil {
+		log.WithError(err).Error("failed in updating volume info to store")
+	}
+
+	return &BlockVolume{
+		hostVolume: volInfo.Name,
+		name:       name,
+		hosts:      resp.Portals,
+		iqn:        resp.IQN,
+		username:   resp.Username,
+		password:   resp.Password,
+		size:       int64(size),
+		ha:         blockVolOpts.Ha,
+	}, nil
+}
+
+// DeleteBlockVolume deletes a gluster block volume of give name
+func (g *GlusterBlock) DeleteBlockVolume(name string, options ...blockprovider.BlockVolOption) error {
+	var (
+		blockVolOpts = &blockprovider.BlockVolumeOptions{}
+		hostVol      string
+	)
+
+	blockVolOpts.ApplyOpts(options...)
+
+	blockVols := g.BlockVolumes()
+
+	for _, blockVol := range blockVols {
+		if blockVol.Name() == name {
+			hostVol = blockVol.HostVolume()
+			break
+		}
+	}
+
+	if hostVol == "" {
+		return errors.New("block volume not found")
+	}
+
+	blockInfo, err := g.client.BlockVolumeInfo(hostVol, name)
+	if err != nil {
+		return err
+	}
+
+	req := &api.BlockVolumeDeleteReq{
+		UnlinkStorage: blockVolOpts.UnlinkStorage,
+		Force:         blockVolOpts.ForceDelete,
+	}
+
+	if err := g.client.DeleteBlockVolume(hostVol, name, req); err != nil {
```

Comment:
question: during the 1st REST call block volume deletion was successful and updating block hosting failed. please add  a TODO to consider this scenario

Reply:
same for create also.